go/vt/vtgate/planbuilder/operators/merging.go

```diff
@@ -0,0 +1,328 @@
+	case a == dual:
+		return m.merge(lhsRoute, rhsRoute, routingB)
+	case b == dual:
+		return m.merge(lhsRoute, rhsRoute, routingA)
+
+		// an unsharded/reference route can be merged with anything going to that keyspace
+	case a == anyShard && sameKeyspace:
+		return m.merge(lhsRoute, rhsRoute, routingB)
+	case b == anyShard && sameKeyspace:
+		return m.merge(lhsRoute, rhsRoute, routingA)
+
+		// None routing can always be merged, as long as we are aiming for the same keyspace
+	case a == none && sameKeyspace:
+		return m.merge(lhsRoute, rhsRoute, routingA)
+	case b == none && sameKeyspace:
+		return m.merge(lhsRoute, rhsRoute, routingB)
+
+		// infoSchema routing is complex, so we handle it in a separate method
+	case a == infoSchema && b == infoSchema:
+		return tryMergeInfoSchemaRoutings(routingA, routingB, m, lhsRoute, rhsRoute)
+
+		// sharded routing is complex, so we handle it in a separate method
+	case a == sharded && b == sharded:
+		return tryMergeShardedRouting(ctx, lhsRoute, rhsRoute, m, joinPredicates)
+
+	default:
+		return nil, nil
+	}
+}
+
+type (
+	merger interface {
+		mergeTables(r1, r2 *ShardedRouting, op1, op2 *Route) (*Route, error)
+		merge(op1, op2 *Route, r Routing) (*Route, error)
+	}
+
+	joinMerger struct {
+		ctx        *plancontext.PlanningContext
+		predicates []sqlparser.Expr
+		innerJoin  bool
+	}
+
+	subQueryMerger struct {
+		ctx  *plancontext.PlanningContext
+		subq *SubQueryInner
+	}
+
+	// mergeDecorator runs the inner merge and also runs the additional function f.
+	mergeDecorator struct {
+		inner merger
+		f     func() error
+	}
+
+	routingType int
+)
+
+const (
+	sharded routingType = iota
+	infoSchema
+	anyShard
+	none
+	dual
+	targeted
+)
+
+func (rt routingType) String() string {
+	switch rt {
+	case sharded:
+		return "sharded"
+	case infoSchema:
+		return "infoSchema"
+	case anyShard:
+		return "anyShard"
+	case none:
+		return "none"
+	case dual:
+		return "dual"
+	case targeted:
+		return "targeted"
+	}
+	panic("switch should be exhaustive")
+}
+
+// getRoutesOrAlternates gets the Routings from each Route. If they are from different keyspaces,
+// we check if this is a table with alternates in other keyspaces that we can use
+func getRoutesOrAlternates(lhsRoute, rhsRoute *Route) (*Route, *Route, Routing, Routing, bool) {
+	routingA := lhsRoute.Routing
+	routingB := rhsRoute.Routing
+	sameKeyspace := routingA.Keyspace() == routingB.Keyspace()
+
+	if !sameKeyspace {
+		if routingA.Keyspace() == nil || routingB.Keyspace() == nil {
+			// if either of these is missing a keyspace, we are not going to be able to find an alternative
+			return lhsRoute, rhsRoute, routingA, routingB, sameKeyspace
+		}
+		refA, ok := routingA.(*AnyShardRouting)
+		if ok {
+			if altARoute := refA.AlternateInKeyspace(routingB.Keyspace()); altARoute != nil {
+				lhsRoute = altARoute
+				routingA = lhsRoute.Routing
+				sameKeyspace = true
+			}
+		}
+	}
+	if !sameKeyspace {
```

Comment:
either the condition of these two `if`s is wrong, or, we can merge the two bocks together

Reply:
by merging the two blocks we would be able to remove a level of indentation in the whole function

Reply:
we are changing `sameKeyspace` inside the first `if` block. In the second we are checking if we still are on different keyspaces or if the first one already fixed it